### PR TITLE
[SPARK-34185][DOCS] Review and fix issues in API docs

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
Compare the 3.1.1 API doc with the latest release version 3.0.1. Fix the following issues:
- Add missing `Since` annotation for new APIs
- Remove the leaking class/object in API doc

### Why are the changes needed?
Fix the issues in the Spark 3.1.1 release API docs.

### Does this PR introduce _any_ user-facing change?
Yes, API doc changes.

### How was this patch tested?
Manually test.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
cc @HyukjinKwon @cloud-fan

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala [hunk 1]
```diff
@@ -31,6 +31,8 @@ case class ExecutorDecommissionInfo(message: String, workerHost: Option[String]
  * State related to decommissioning that is kept by the TaskSchedulerImpl. This state is derived
  * from the info message above but it is kept distinct to allow the state to evolve independently
  * from the message.
+ *
+ * @since 3.1.0
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:35]: @holdenk @Ngone51 do we mean to expose this class to end-users?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -1268,7 +1268,7 @@ private[hive] object HiveClientImpl extends Logging {
   }
 }
 
-case object HiveVoidType extends DataType {
+private[hive] case object HiveVoidType extends DataType {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:1271]: hmmm, I thought `org.apache.spark.sql.hive` is already a private package, no? cc @viirya @dongjoon-hyun

Comment (viirya) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:1271]: I think so.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryPage.scala [hunk 3]
```diff
@@ -206,12 +206,13 @@ class StreamingQueryPagedTable(
...
 }
 
-case class StructuredStreamingRow(
+private[ui] case class StructuredStreamingRow(
     duration: Long,
     avgInput: Double,
     avgProcess: Double,
     streamingUIData: StreamingQueryUIData)
 
+private[ui]
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryPage.scala:215]: nit:
shall we make it 
```
private[ui] class StreamingQueryDataSource(
    uiData: Seq[StreamingQueryUIData],
    ...
```

Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryPage.scala:215]: Make sense, will fix in the next commit.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:1271]: Nah, it's exposed. https://spark.apache.org/docs/3.0.1/api/java/org/apache/spark/sql/hive/package-summary.html. We should better explicitly make it private

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala (hunk 1 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:35]: I don't think so. It's only used internally.

Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:35]: Copy, will add @Private tag for it.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 2 above)
Comment (xuanyuanking) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:1271]: Got it, will fix it in the next commit.

File: project/SparkBuild.scala [hunk 4]
```diff
@@ -910,7 +910,7 @@ object Unidoc {
       .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/catalyst")))
       .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/execution")))
       .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/internal")))
-      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/hive/test")))
+      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/hive")))
```
Comment (HyukjinKwon) [inline project/SparkBuild.scala:913]: @xuanyuanking, can we avoid doing this here? Making private sounds correct at this moment. Hive module at least has one API exposed `HiveSessionStateBuilder`.

Comment (HyukjinKwon) [reply project/SparkBuild.scala:913]: Also, we should document that this is a private module at https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/package.scala. We could do it separately in another PR.

**Review by HyukjinKwon — APPROVED** [review]

File: project/SparkBuild.scala (hunk 4 above)
Comment (cloud-fan) [reply project/SparkBuild.scala:913]: I don't think `HiveSessionStateBuilder` is a public API. It's base class, `BaseSessionStateBuilder`, is in `org.apache.spark.sql.internal`.

Comment (HyukjinKwon) [reply project/SparkBuild.scala:913]: Then we should remove `@Unstable`. Can we do this in a separate PR?

Comment (cloud-fan) [reply project/SparkBuild.scala:913]: yea we should do it in a separated PR.

Comment (xuanyuanking) [reply project/SparkBuild.scala:913]: Copy that, will do it in a separated PR.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 5]
```diff
@@ -90,7 +90,7 @@ private[storage] class FallbackStorage(conf: SparkConf) extends Logging {
   }
 }
 
-class NoopRpcEndpointRef(conf: SparkConf) extends RpcEndpointRef(conf) {
+private[storage] class NoopRpcEndpointRef(conf: SparkConf) extends RpcEndpointRef(conf) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:93]: Thank you for fixing this.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:1271]: Thank you for pinging me. +1 for making it private.

**Review by dongjoon-hyun — APPROVED** [review]

**Review by viirya — APPROVED** [review]

File: project/SparkBuild.scala (hunk 4 above)
Comment (HyukjinKwon) [inline project/SparkBuild.scala:913]:
```suggestion
      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/hive/test")))
```

Comment (HyukjinKwon) [reply project/SparkBuild.scala:913]: Let's do this change in a separate PR that makes `hive` a private package

**Comment by HyukjinKwon** [comment]
I will merge this one for Spark 3.1.1 RC preparation. Let's handle the private hive module one separately in another PR.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1.

**Comment by xuanyuanking** [comment]
Thanks all! Will handle the private hive module today.

File: project/SparkBuild.scala (hunk 4 above)
Comment (xuanyuanking) [reply project/SparkBuild.scala]: Done in #31321

**Pull request closed without merging**